libV2/redis.js

```diff
@@ -0,0 +1,160 @@
+const EventEmitter = require('events');
+const { callbackify, promisify } = require('util');
+const IORedis = require('ioredis');
+const uuid = require('uuid');
+
+const errors = require('./errors');
+const { LoggerContext, asyncOrCallback } = require('./utils');
+
+const moduleLogger = new LoggerContext({
+    module: 'redis',
+});
+
+const COMMAND_TIMEOUT = 10000;
+const CONNECTION_TIMEOUT = 30000;
+
+
+class RedisClient extends EventEmitter {
+    constructor(options) {
+        super();
+        this._redisOptions = options;
+        this._redis = null;
+        // Controls the use of additional command timeouts
+        // Only use if connecting to a sentinel cluster
+        this._useTimeouts = options.sentinels !== undefined;
+        this._inFlightTimeouts = this._useTimeouts ? {} : null;
+        this._runningRedisProbe = null;
+        this._isConnected = false;
+        this._isReady = false;
+    }
+
+    connect(callback) {
+        this._initClient(false);
+        if (callback) {
+            callback();
+        }
+    }
+
+    disconnect(callback) {
+        return asyncOrCallback(async () => {
+            if (this._useTimeouts) {
+                Object.values(this._inFlightTimeouts)
+                    .forEach(clearTimeout);
+            }
+            await this._redis.quit();
+        }, callback);
+    }
+
+    get isReady() {
+        return this._isConnected && this._isReady;
+    }
+
+    _initClient(startProbe = true) {
+        moduleLogger.debug('initializing redis client');
+        if (this._redis !== null) {
+            this._redis.off('connect', this._onConnect);
+            this._redis.off('ready', this._onReady);
+            this._redis.off('error', this._onError);
+        }
+        this._isConnected = false;
+        this._isReady = false;
+        this._redis = new IORedis(this._redisOptions);
+        this._redis.on('connect', this._onConnect.bind(this));
+        this._redis.on('ready', this._onReady.bind(this));
+        this._redis.on('error', this._onError.bind(this));
```

Comment:
Just curious to learn. What additional functionality does .bind bring?

Reply:
Without `.bind`, when the handlers are called they would lose their reference to the correct `this`. Normally I would just wrap it in an anonymous function  like `() => this._onReady()` however doing so would would mean that the anonymous function is registered rather than the instance method which would prevent me from using `.off` to remove them.